pytensor/tensor/rewriting/linalg.py

```diff
@@ -539,3 +540,91 @@ def svd_uv_merge(fgraph, node):
+    # If the current op is solve, we check for b. If b is an identity matrix (Eye), we can return True
+    solve_inputs = node.inputs
+    eye_node = solve_inputs[1].owner
+
+    # We check for b = Eye and also make sure that if it was an Eye, then k = 0 (1's are only across the main diagonal)
+    if not (eye_node and isinstance(eye_node.op, Eye)):
+        return False
+
+    if (
+        isinstance(eye_node.inputs[-1], Constant)
+        and eye_node.inputs[-1].data.item() != 0
+    ):
+        return False
+    return True
+
+
+@register_canonicalize
+@register_stabilize
+@node_rewriter([Blockwise])
+def rewrite_inv_inv(fgraph, node):
+    """
+    This rewrite takes advantage of the fact that if there are two consecutive inverse operations (inv(inv(input))), we get back our original input without having to compute inverse once.
+
+    Here, we check for direct inverse operations (inv/pinv) and also solve operations (solve/solve_triangular) in the case when b = Eye. This allows any combination of these "inverse" nodes to be simply rewritten.
+
+    Parameters
+    ----------
+    fgraph: FunctionGraph
+        Function graph being optimized
+    node: Apply
+        Node of the function graph to be optimized
+
+    Returns
+    -------
+    list of Variable, optional
+        List of optimized variables, or None if no optimization was performed
+    """
+    valid_inverses = (MatrixInverse, MatrixPinv, Solve, SolveTriangular)
+    valid_solves = (Solve, SolveTriangular)
+    # Check if its a valid inverse operation (either inv/pinv or if its solve, then b = eye)
+    # In case the outer operation is an inverse, it directly goes to the next step of finding inner operation
+    # If the outer operation is a solve op with b = Eye, it treats it as inverse and finds the inner operation
+    # If the outer operation is not an inverse (neither inv nor solve with eye), we do not apply this rewrite
+    if not isinstance(node.op.core_op, valid_inverses):
+        return None
+
+    if isinstance(node.op.core_op, valid_solves) and not _find_solve_with_eye(node):
+        return None
+
+    potential_inner_inv = node.inputs[0].owner
+    if potential_inner_inv is None or potential_inner_inv.op is None:
+        return None
+
+    # Check if inner op is blockwise and and possible inv
+    if not (
+        potential_inner_inv
+        and isinstance(potential_inner_inv.op, Blockwise)
+        and isinstance(node.op.core_op, valid_inverses)
+    ):
+        return None
+
+    # Similar to the check for outer operation, we now run the same checks for the inner op.
+    # If its an inverse or solve with eye, we apply the rewrite. Otherwise, we return None.
+    if not (
+        isinstance(potential_inner_inv.op, Blockwise)
+        and isinstance(potential_inner_inv.op.core_op, valid_inverses)
+    ):
+        return None
+
+    if isinstance(
+        potential_inner_inv.op.core_op, valid_solves
+    ) and not _find_solve_with_eye(potential_inner_inv):
```

Comment:
Instead of looking for solve with eye in all rewrites that are interested in matrix inversions, we can canonicalize such solves as matrix inverses in a separate rewrite?

Reply:
I think rewriting `solve(x, eye) -> inv(x)` is a good idea following the discussion with Adrien yesterday. 

If we want to do that, I propose we remove everything to do with `solve` and `solve_triangular` in this PR, and open another issue later to deal with the canonicalization.

Reply:
so for now this should also only deal with inv and pinv?
there is a function here called _find_solve_with_eye so i should get rid of that too?

Reply:
Yeah you can pull it out of this PR, but I suggest you save it in a new branch so we can re-use it in this proposed `solve(x, eye) -> inv(x)` rewrite